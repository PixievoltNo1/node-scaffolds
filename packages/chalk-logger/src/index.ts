import type { Mutable } from '@guanghechen/utility-types'
import { partOptionsFromArgs } from './commander'
import type { Level } from './level'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export * from './color'
export * from './commander'
export * from './level'
export * from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = partOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flags: { ...options?.flags, ...optionsFromArgs.flags },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's level
   * @param level
   */
  public setLevel(level: Level | null | undefined): void {
    if (level == null) return
    const self = this as Mutable<this>
    self.level = level
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}

export default ChalkLogger
