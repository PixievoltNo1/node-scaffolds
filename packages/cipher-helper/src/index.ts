export * from './_base'
export * from './aes'
export * from './catalog'
export * from './types/catalog'
export * from './types/cipher-helper'
export * from './util/buffer'
export * from './util/key'
