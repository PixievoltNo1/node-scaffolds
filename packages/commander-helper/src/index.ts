export * from './command'
export * from './fs'
export * from './git'
export * from './merge'
export * from './option'
export * from './yarn'
