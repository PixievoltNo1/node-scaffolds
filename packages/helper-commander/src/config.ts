import { ensureCriticalFilepathExistsSync } from '@guanghechen/helper-file'
import fs from 'fs-extra'
import yaml from 'js-yaml'
import path from 'path'

/**
 * Load configuration file with format .json / .yml / .yaml
 *
 * @param filepath
 * @param encoding
 * @returns
 */
export async function loadJsonOrYaml(
  filepath: string,
  encoding: BufferEncoding = 'utf8',
): Promise<unknown | never> {
  ensureCriticalFilepathExistsSync(filepath)
  const loadContent = (): Promise<string> => fs.readFile(filepath, encoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json': {
      const content: string = await loadContent()
      result = JSON.parse(content)
      break
    }
    case '.yml':
    case '.yaml': {
      const content: string = await loadContent()
      result = yaml.load(content, { filename: filepath, json: true })
      break
    }
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}

/**
 * Load configuration file with format .json / .yml / .yaml  (synchronizing)
 *
 * @param filepath
 * @param encoding
 * @returns
 */
export function loadJsonOrYamlSync(
  filepath: string,
  encoding: BufferEncoding = 'utf8',
): unknown | never {
  ensureCriticalFilepathExistsSync(filepath)
  const loadContent = (): string => fs.readFileSync(filepath, encoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json': {
      const content: string = loadContent()
      result = JSON.parse(content)
      break
    }
    case '.yml':
    case '.yaml': {
      const content: string = loadContent()
      result = yaml.load(content, { filename: filepath, json: true })
      break
    }
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}
