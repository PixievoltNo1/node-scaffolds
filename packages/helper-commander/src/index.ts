export * from './command'
export * from './config'
export * from './git'
export * from './merge'
export * from './option'
export * from './stdin'
export * from './strip-ansi'
export * from './yarn'
