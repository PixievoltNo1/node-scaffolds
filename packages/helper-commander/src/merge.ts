import { isNotEmptyArray, isNotEmptyObject } from '@guanghechen/helper-is'

/**
 * The strategy for merging options
 *
 * @param prevOption   previous option item
 * @param nextOption   next option item
 * @returns option item merged prevOption and nextOption
 */
export type IMergeStrategy<T = unknown> = (prevOption: T, nextOption: T | null) => T

/**
 * Default merging strategy
 */
export const defaultMergeStrategies = {
  /**
   * Replace the previous item with new one if it's neither null nor undefined,
   * otherwise, retain the previous one
   */
  replace: function (prevOption: unknown, nextOption: unknown | null): unknown {
    return nextOption == null ? prevOption : nextOption
  } as IMergeStrategy,
  /**
   * Retain the previous item and ignore the new one if the previous one is
   * neither null nor undefined, otherwise, use the new one instead
   */
  retain: function (prevOption: unknown): unknown {
    return prevOption
  } as IMergeStrategy,
  /**
   * Merge arrays, like Array.prototype.concat
   */
  concat: function (prevOption: unknown[], nextOption: unknown[] | null): unknown[] {
    if (isNotEmptyArray(nextOption)) {
      return prevOption.length > 0 ? [...prevOption, ...nextOption] : nextOption
    }
    return prevOption
  } as IMergeStrategy<unknown[]>,
  /**
   * Merge objects, like Object.prototype.assign
   */
  assign: function (
    prevOption: Record<string, unknown>,
    nextOption: Record<string, unknown> | null,
  ): Record<string, unknown> {
    if (isNotEmptyObject(nextOption)) {
      return { ...prevOption, ...nextOption }
    }
    return prevOption
  } as IMergeStrategy<Record<string, unknown>>,
}

/**
 * Merge options
 * @param options
 * @param strategies
 */
export function merge<O extends Record<string, unknown>>(
  options: O[],
  strategies: Partial<Record<keyof O, IMergeStrategy<any>>> = {},
  defaultStrategy: IMergeStrategy = defaultMergeStrategies.replace,
): O {
  const result = {} as unknown as O
  for (const option of options) {
    for (const key of Object.keys(option)) {
      const strategy: IMergeStrategy = strategies[key] || defaultStrategy
      if (result[key] != null) {
        result[key as keyof O] = strategy(result[key], option[key]) as O[keyof O]
      } else {
        result[key as keyof O] = option[key] as O[keyof O]
      }
    }
  }
  return result
}
