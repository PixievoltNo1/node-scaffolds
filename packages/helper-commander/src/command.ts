import { registerCommanderOptions } from '@guanghechen/chalk-logger'
import type { Option, OptionValues } from 'commander'
import commander from 'commander'
import type { ICommandConfigurationOptions } from './option'

/**
 * Callback for handling the command
 *
 * @param args    command arguments
 * @param options command options
 * @param extra   extra args (neither declared command arguments nor command options)
 */
export type ICommandActionCallback<T extends ICommandConfigurationOptions> = (
  args: string[],
  options: T,
  extra: string[],
) => void | Promise<void> | never

export class Command extends commander.Command {
  // add missing type declarations
  public options: Option[] | undefined
  protected _actionResults: unknown[] | undefined
  protected _storeOptionsAsProperties: boolean | undefined
  protected _optionValues: Record<string, unknown> | undefined
  protected _actionHandler: ((args: string[]) => void) | null | undefined
  protected _versionOptionName: string | undefined
  protected _version: string | undefined

  /**
   * Register callback `fn` for the command.
   *
   * Examples:
   *
   *      program
   *        .command('help')
   *        .description('display verbose help')
   *        .action(function() {
   *           // output help here
   *        });
   *
   * @param {Function} fn
   * @return {Command} `this` command for chaining
   * @api public
   */
  public override action<T extends ICommandConfigurationOptions>(
    fn: ICommandActionCallback<T>,
  ): this {
    const self = this

    const listener = (args: string[]): void => {
      // eslint-disable-next-line max-len
      // The .action callback takes an extra parameter which is the command or options.
      const expectedArgsCount = self.args.length

      // const actionArgs: (string | Record<string, unknown> | string[])[] = [
      const actionArgs: [string[], T, string[]] = [
        // Command arguments
        args.slice(0, expectedArgsCount),

        // Command options
        self.opts() as T,

        // Extra arguments so available too.
        args.slice(expectedArgsCount),
      ]

      const actionResult = fn.apply(self, actionArgs)

      // Remember result in case it is async.
      // Assume parseAsync getting called on root.
      let rootCommand: Command = self
      while (rootCommand.parent != null) {
        rootCommand = rootCommand.parent as Command
      }
      if (rootCommand._actionResults != null) {
        rootCommand._actionResults.push(actionResult)
      }
    }

    self._actionHandler = listener
    return self
  }

  public override opts<T extends OptionValues>(): T {
    const self = this
    const nodes: Command[] = [self]
    for (let parent = self.parent; parent != null; parent = parent.parent) {
      nodes.push(parent as Command)
    }

    const options: Record<string, unknown> = {}
    for (let i = nodes.length - 1; i >= 0; --i) {
      const o = nodes[i]
      if (o._storeOptionsAsProperties) {
        for (const option of o.options!) {
          const key = option.attributeName()
          options[key] = key === o._versionOptionName ? o._version : o[key]
        }
      } else {
        for (const key of Object.getOwnPropertyNames(o._optionValues)) {
          options[key] = o._optionValues![key]
        }
      }
    }

    return options as unknown as T
  }

  // override
  public override addCommand(command: Command, opts?: commander.CommandOptions): this {
    super.addCommand(command, opts)
    return this
  }
}

export { commander }

/**
 * Create top command
 * @param commandName
 * @param version
 */
export function createTopCommand(commandName: string, version: string): Command {
  const program = new Command()

  program.storeOptionsAsProperties(false).version(version).name(commandName)

  registerCommanderOptions(program)

  program
    .option(
      '-c, --config-path <configFilepath>',
      'config filepaths',
      (val, acc: string[]) => acc.concat(val),
      [],
    )
    .option('--parastic-config-path <parasticConfigFilepath>', 'parastic config filepath')
    .option('--parastic-config-entry <parasticConfigFilepath>', 'parastic config filepath')

  return program
}

// Process sub-command
export type ISubCommandProcessor<O extends ICommandConfigurationOptions, V = void> = (
  options: O,
) => V | Promise<V>

// Create sub-command
export type ISubCommandCreator<O extends ICommandConfigurationOptions, V = void> = (
  handle?: ISubCommandProcessor<O, V>,
  commandName?: string,
  aliases?: string[],
) => Command

// Mount sub-command
export type ISubCommandMounter = (parentCommand: Command, opts?: commander.CommandOptions) => void

// Execute sub-command
export type ISubCommandExecutor<V = void> = (parentCommand: Command, args: string[]) => Promise<V>

/**
 * Create sub-command mounter
 *
 * @param create  sub command creator
 * @param handle  sub command processor
 */
export function createSubCommandMounter<O extends ICommandConfigurationOptions, V = void>(
  create: ISubCommandCreator<O, V>,
  handle: ISubCommandProcessor<O, V>,
): ISubCommandMounter {
  return (program: Command, opts?: commander.CommandOptions): void => {
    const command = create(handle)
    program.addCommand(command, opts)
  }
}

/**
 * Create sub-command executor
 *
 * @param create        sub-command creator
 * @param handle        sub-command processor
 * @param commandName   sub-command name
 * @param aliases       sub-command aliases
 */
export function createSubCommandExecutor<O extends ICommandConfigurationOptions, V = void>(
  create: ISubCommandCreator<O, V>,
  handle: ISubCommandProcessor<O, V>,
  commandName?: string,
  aliases?: string[],
): ISubCommandExecutor<V> {
  return (parentCommand: Command, args: string[]): Promise<V> => {
    return new Promise(resolve => {
      const wrappedHandler = async (options: O): Promise<V> => {
        const result = await handle(options)
        resolve(result)
        return result
      }

      const command = create(wrappedHandler, commandName, aliases)
      parentCommand.addCommand(command)
      parentCommand.parse(args)
    })
  }
}

// Process main command.
export type IMainCommandProcessor<O extends ICommandConfigurationOptions, V = void> = (
  options: O,
) => V | Promise<V>

// Create main command
export type IMainCommandCreator<O extends ICommandConfigurationOptions, V = void> = (
  handle?: IMainCommandProcessor<O, V>,
) => Command

// Mount main command
export type IMainCommandMounter = (parentCommand: Command, opts?: commander.CommandOptions) => void

// Execute main command.
export type IMainCommandExecutor<V = void> = (args: string[]) => Promise<V>

/**
 * Create main command mounter
 *
 * @param create  main command creator
 * @param handle  main command processor
 */
export function createMainCommandMounter<O extends ICommandConfigurationOptions, V = void>(
  create: IMainCommandCreator<O, V>,
  handle: IMainCommandProcessor<O, V>,
): IMainCommandMounter {
  return (program: Command, opts?: commander.CommandOptions): void => {
    const command = create(handle)
    if (command.name().length <= 0) {
      command.name('__main__')
    }
    program.addCommand(command, { ...opts, isDefault: true })
  }
}

/**
 * Create main command executor
 *
 * @param create  main command creator
 * @param handle  main command processor
 */
export function createMainCommandExecutor<O extends ICommandConfigurationOptions, V = void>(
  create: IMainCommandCreator<O, V>,
  handle: IMainCommandProcessor<O, V>,
): IMainCommandExecutor<V> {
  return (args: string[]): Promise<V> => {
    return new Promise(resolve => {
      const wrappedHandler = async (options: O): Promise<V> => {
        const result = await handle(options)
        resolve(result)
        return result
      }

      const command = create(wrappedHandler)
      command.parse(args)
    })
  }
}
