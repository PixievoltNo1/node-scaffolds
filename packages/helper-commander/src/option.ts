import { Level } from '@guanghechen/chalk-logger'
import type { ChalkLogger } from '@guanghechen/chalk-logger'
import { isNonBlankString, isNotEmptyArray, isNotEmptyObject } from '@guanghechen/helper-is'
import { cover, coverString } from '@guanghechen/helper-option'
import { absoluteOfWorkspace, locateNearestFilepath } from '@guanghechen/helper-path'
import path from 'path'
import { loadJsonOrYamlSync } from './fs'
import type { IMergeStrategy } from './merge'
import { defaultMergeStrategies, merge } from './merge'

export interface ICommandConfigurationFlatOpts {
  /**
   * Path of currently executing command.
   */
  readonly cwd: string

  /**
   * Working directory.
   */
  readonly workspace: string

  /**
   * Filepath of configs, only *.yml, *.yaml and *.json are supported.
   * Each configuration file can specify the same options, the configuration
   * file specified later can override the configuration specified previous.
   * @default []
   */
  readonly configPath?: string[]

  /**
   * Filepath of parastic config,
   */
  readonly parasticConfigPath?: string | null

  /**
   * The entry key of options in the parasitic configuration file
   */
  readonly parasticConfigEntry?: string | null
}

export interface ICommandConfigurationOptions extends Record<string, unknown> {
  /**
   * log level
   * @default undefined
   */
  readonly logLevel?: 'debug' | 'verbose' | 'info' | 'warn' | 'error' | string

  /**
   * Filepath of configs, only *.yml, *.yaml and *.json are supported.
   * Each configuration file can specify the same options, the configuration
   * file specified later can override the configuration specified previous.
   * @default []
   */
  readonly configPath?: string[]

  /**
   * Filepath of parastic config,
   */
  readonly parasticConfigPath?: string | null

  /**
   * The entry key of options in the parasitic configuration file
   */
  readonly parasticConfigEntry?: string | null
}

export interface ICommandConfiguration<Options extends ICommandConfigurationOptions> {
  /**
   * Global options shared by all sub-commands
   */
  __globalOptions__: Options

  /**
   * Sub-command specific options
   */
  [subCommand: string]: Options
}

/**
 * Flat defaultOptions with configs from package.json
 */
export function flatOptionsFromConfiguration<C extends ICommandConfigurationOptions>(
  defaultOptions: C,
  flatOpts: ICommandConfigurationFlatOpts,
  subCommandName: string | false,
  strategies: Partial<Record<keyof C, IMergeStrategy>> = {},
): C {
  let resolvedConfig = {} as unknown as ICommandConfiguration<C>

  // load configs
  if (isNotEmptyArray(flatOpts.configPath)) {
    const configs: Array<ICommandConfiguration<C>> = []
    for (const filepath of flatOpts.configPath) {
      const config = loadJsonOrYamlSync(filepath) as ICommandConfiguration<C>
      configs.push(config)
    }
    resolvedConfig = merge<ICommandConfiguration<C>>(configs, {}, defaultMergeStrategies.replace)
  } else {
    // otherwise, load from parastic config
    if (
      isNonBlankString(flatOpts.parasticConfigPath) &&
      isNonBlankString(flatOpts.parasticConfigEntry)
    ) {
      const config = loadJsonOrYamlSync(flatOpts.parasticConfigPath) as any
      resolvedConfig = (config[flatOpts.parasticConfigEntry] as ICommandConfiguration<C>) || {}
    }
  }

  let result: C = defaultOptions
  if (subCommandName === false) {
    result = merge<C>(
      [result, resolvedConfig as unknown as C],
      strategies,
      defaultMergeStrategies.replace,
    )
  } else {
    // merge globalOptions
    if (isNotEmptyObject(resolvedConfig.__globalOptions__)) {
      result = merge<C>(
        [result, resolvedConfig.__globalOptions__],
        strategies,
        defaultMergeStrategies.replace,
      )
    }

    // merge specified sub-command option
    if (isNonBlankString(subCommandName) && typeof resolvedConfig[subCommandName] == 'object') {
      result = merge<C>(
        [result, resolvedConfig[subCommandName]],
        strategies,
        defaultMergeStrategies.replace,
      )
    }
  }

  return result
}

/**
 * Resolve CommandConfigurationOptions
 *
 * @param logger
 * @param commandName
 * @param subCommandName
 * @param workspaceDir
 * @param defaultOptions
 * @param options
 * @param strategies
 */
export function resolveCommandConfigurationOptions<
  C extends Partial<ICommandConfigurationOptions>,
  D extends ICommandConfigurationOptions,
>(
  logger: ChalkLogger,
  commandName: string,
  subCommandName: string | false,
  workspaceDir: string,
  defaultOptions: D,
  options: C,
  strategies: Partial<Record<keyof D, IMergeStrategy>> = {},
): D & ICommandConfigurationFlatOpts {
  const cwd: string = path.resolve()
  const workspace: string = path.resolve(cwd, workspaceDir)
  const configPath: string[] = (options.configPath ?? defaultOptions.configPath ?? []).map(
    (p: string) => absoluteOfWorkspace(workspace, p),
  )
  const parasticConfigPath: string | null | undefined = cover<string | null>(
    () => locateNearestFilepath(workspace, 'package.json'),
    options.parasticConfigPath,
  )
  const parasticConfigEntry: string = coverString(commandName, options.parasticConfigEntry)
  const flatOpts: ICommandConfigurationFlatOpts = {
    cwd,
    workspace,
    configPath,
    parasticConfigPath,
    parasticConfigEntry,
  }

  const resolvedOptions = flatOptionsFromConfiguration<D>(
    defaultOptions,
    flatOpts,
    subCommandName,
    strategies,
  )

  // reset log-level
  const logLevel = cover<string | undefined>(resolvedOptions.logLevel, options.logLevel)
  if (logLevel != null) {
    const level = Level.valueOf(logLevel)
    if (level != null) logger.setLevel(level)
  }

  logger.debug('cwd:', flatOpts.cwd)
  logger.debug('workspace:', flatOpts.workspace)
  logger.debug('configPath:', flatOpts.configPath)
  logger.debug('parasticConfigPath:', flatOpts.parasticConfigPath)
  logger.debug('parasticConfigEntry:', flatOpts.parasticConfigEntry)

  return {
    ...resolvedOptions,
    logLevel,
    cwd,
    workspace,
    configPath,
    parasticConfigPath,
    parasticConfigEntry,
  }
}
