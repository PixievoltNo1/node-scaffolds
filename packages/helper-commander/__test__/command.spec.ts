import ChalkLogger from '@guanghechen/chalk-logger'
import { isNonBlankString } from '@guanghechen/helper-is'
import { createLoggerMock } from '@guanghechen/helper-jest'
import { convertToBoolean, cover } from '@guanghechen/helper-option'
import { absoluteOfWorkspace } from '@guanghechen/helper-path'
import { desensitize } from 'jest.helper'
import path from 'path'
import type { Command, ICommandConfigurationFlatOpts, ICommandConfigurationOptions } from '../src'
import { createTopCommand, resolveCommandConfigurationOptions } from '../src'

describe('command', () => {
  describe('no sub-command', () => {
    test('with arguments', async () => {
      const argv: string[] = [
        'node',
        'guanghechen',
        'source content waw waw waw',
        'this arg should be ignored',
        '--log-level=debug',
        '--log-flag=no-date',
        '--log-flag=no-color',
        '-f',
        '--no-silence',
      ]

      const logger = new ChalkLogger({ name: 'ghc', flags: { colorful: false, date: false } }, argv)
      const mock = createLoggerMock({ logger, desensitize })
      const result0 = await getCommand(argv, logger)
      expect(desensitize(result0.args)).toMatchSnapshot('args')
      expect(desensitize(result0.options)).toMatchSnapshot('options')
      expect(desensitize(result0.resolvedOptions)).toMatchSnapshot('resolvedOptions')
      expect(mock.getIndiscriminateAll()).toMatchSnapshot('logs')
      mock.restore()
    })

    test('without arguments', async () => {
      const argv: string[] = [
        'node',
        'guanghechen',
        '--log-level=debug',
        '--log-flag=no-date',
        '--log-flag=no-color',
        '-f',
        '--no-silence',
      ]

      const logger = new ChalkLogger({ name: 'ghc', flags: { colorful: false, date: false } }, argv)
      const mock = createLoggerMock({ logger, desensitize })
      const result0 = await getCommand(argv, logger)
      expect(desensitize(result0.args)).toMatchSnapshot('args')
      expect(desensitize(result0.options)).toMatchSnapshot('options')
      expect(desensitize(result0.resolvedOptions)).toMatchSnapshot('resolvedOptions')
      expect(mock.getIndiscriminateAll()).toMatchSnapshot('logs')
      mock.restore()
    })
  })
})

async function getCommand(
  argv: string[],
  logger: ChalkLogger,
): Promise<{
  args: string[]
  options: IGlobalCommandOptions
  resolvedOptions: IGlobalCommandOptions & ICommandConfigurationFlatOpts
  program: Command
}> {
  type ICommandData = Awaited<ReturnType<typeof getCommand>>

  const __defaultGlobalCommandOptions: IGlobalCommandOptions = {
    encoding: 'utf8',
    input: undefined,
    output: undefined,
    force: false,
    silence: true,
  }

  const program = createTopCommand('guanghechen', '0.0.0')
  const result: ICommandData = await new Promise<ICommandData>((resolve, reject) => {
    program
      .argument('[source content')
      .option('-e, --encoding <encoding>', 'Encoding of content from stdin or file.')
      .option('-i, --input <filepath>', 'Copy the data from <filepath> to the system clipboard.')
      .option(
        '-o, --output <filepath>',
        'Write the data from the system clipboard into <filepath>.',
      )
      .option('-f, --force', 'Overwrite the <filepath> without confirmation.')
      .option('-S, --no-silence', 'print info level logs.')
      .action(function (args: string[], options: IGlobalCommandOptions): void {
        try {
          const resolvedOptions = resolveGlobalCommandOptions(
            '@guanghechen/helper-commander--demo',
            '',
            __defaultGlobalCommandOptions,
            path.resolve(),
            options,
            logger,
          )
          resolve({ args, options, resolvedOptions, program })
        } catch (error) {
          reject(error)
        }
      })
      .parse(argv)
  })
  return result
}

interface IGlobalCommandOptions extends ICommandConfigurationOptions {
  encoding: BufferEncoding
  input?: string
  output?: string
  force: boolean
  silence: boolean
}

function resolveGlobalCommandOptions<O extends object>(
  commandName: string,
  subCommandName: string | false,
  defaultOptions: O & IGlobalCommandOptions,
  workspaceDir: string,
  options: O & IGlobalCommandOptions,
  logger: ChalkLogger,
): O & IGlobalCommandOptions & ICommandConfigurationFlatOpts {
  type R = O & IGlobalCommandOptions & ICommandConfigurationFlatOpts
  const resolvedDefaultOptions: R = resolveCommandConfigurationOptions<O & IGlobalCommandOptions>(
    logger,
    commandName,
    subCommandName,
    workspaceDir,
    defaultOptions,
    options,
  )

  // Resolve `encoding`.
  const encoding: string = cover<string>(
    resolvedDefaultOptions.encoding,
    options.encoding,
    isNonBlankString,
  )
  logger.debug('encoding:', encoding)

  // Resolve `input`.
  const _input: string | undefined = cover<string | undefined>(
    resolvedDefaultOptions.input,
    options.input,
    isNonBlankString,
  )
  const input: string | undefined = _input ? absoluteOfWorkspace(workspaceDir, _input) : undefined
  logger.debug('input:', input)

  // Resolve `output`.
  const _output: string | undefined = cover<string | undefined>(
    resolvedDefaultOptions.output,
    options.output,
    isNonBlankString,
  )
  const output: string | undefined = _output
    ? absoluteOfWorkspace(workspaceDir, _output)
    : undefined
  logger.debug('output:', output)

  // Resolve `force`.
  const force: boolean = cover<boolean>(
    resolvedDefaultOptions.force,
    convertToBoolean(options.force),
  )
  logger.debug('force:', force)

  // Resolve `silence`.
  const silence: boolean = cover<boolean>(
    resolvedDefaultOptions.silence,
    convertToBoolean(options.silence),
  )
  logger.debug('silence:', silence)

  return {
    ...resolvedDefaultOptions,
    encoding,
    input,
    output,
    force,
    silence,
  }
}
