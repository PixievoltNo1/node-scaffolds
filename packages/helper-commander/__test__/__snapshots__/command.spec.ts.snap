// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command no sub-command with arguments: args 1`] = `
[
  "source content waw waw waw",
]
`;

exports[`command no sub-command with arguments: logs 1`] = `
[
  [
    "[debug ghc] cwd: <$WORKSPACE$>/packages/helper-commander
",
  ],
  [
    "[debug ghc] workspace: <$WORKSPACE$>/packages/helper-commander
",
  ],
  [
    "[debug ghc] configPath: []
",
  ],
  [
    "[debug ghc] parasticConfigPath: <$WORKSPACE$>/packages/helper-commander/package.json
",
  ],
  [
    "[debug ghc] parasticConfigEntry: @guanghechen/helper-commander--demo
",
  ],
  [
    "[debug ghc] logLevel: debug
",
  ],
  [
    "[debug ghc] encoding: utf8
",
  ],
  [
    "[debug ghc] input: undefined
",
  ],
  [
    "[debug ghc] output: undefined
",
  ],
  [
    "[debug ghc] force: true
",
  ],
  [
    "[debug ghc] silence: false
",
  ],
]
`;

exports[`command no sub-command with arguments: options 1`] = `
{
  "configPath": [],
  "force": true,
  "logFlag": [
    "no-date",
    "no-color",
  ],
  "logLevel": "debug",
  "silence": false,
}
`;

exports[`command no sub-command with arguments: resolvedOptions 1`] = `
{
  "configPath": [],
  "cwd": "<$WORKSPACE$>/packages/helper-commander",
  "encoding": "utf8",
  "force": true,
  "input": undefined,
  "logLevel": "debug",
  "output": undefined,
  "parasticConfigEntry": "@guanghechen/helper-commander--demo",
  "parasticConfigPath": "<$WORKSPACE$>/packages/helper-commander/package.json",
  "silence": false,
  "workspace": "<$WORKSPACE$>/packages/helper-commander",
}
`;

exports[`command no sub-command without arguments: args 1`] = `[]`;

exports[`command no sub-command without arguments: logs 1`] = `
[
  [
    "[debug ghc] cwd: <$WORKSPACE$>/packages/helper-commander
",
  ],
  [
    "[debug ghc] workspace: <$WORKSPACE$>/packages/helper-commander
",
  ],
  [
    "[debug ghc] configPath: []
",
  ],
  [
    "[debug ghc] parasticConfigPath: <$WORKSPACE$>/packages/helper-commander/package.json
",
  ],
  [
    "[debug ghc] parasticConfigEntry: @guanghechen/helper-commander--demo
",
  ],
  [
    "[debug ghc] logLevel: debug
",
  ],
  [
    "[debug ghc] encoding: utf8
",
  ],
  [
    "[debug ghc] input: undefined
",
  ],
  [
    "[debug ghc] output: undefined
",
  ],
  [
    "[debug ghc] force: true
",
  ],
  [
    "[debug ghc] silence: false
",
  ],
]
`;

exports[`command no sub-command without arguments: options 1`] = `
{
  "configPath": [],
  "force": true,
  "logFlag": [
    "no-date",
    "no-color",
  ],
  "logLevel": "debug",
  "silence": false,
}
`;

exports[`command no sub-command without arguments: resolvedOptions 1`] = `
{
  "configPath": [],
  "cwd": "<$WORKSPACE$>/packages/helper-commander",
  "encoding": "utf8",
  "force": true,
  "input": undefined,
  "logLevel": "debug",
  "output": undefined,
  "parasticConfigEntry": "@guanghechen/helper-commander--demo",
  "parasticConfigPath": "<$WORKSPACE$>/packages/helper-commander/package.json",
  "silence": false,
  "workspace": "<$WORKSPACE$>/packages/helper-commander",
}
`;
